draft-gpew-priv-ppm.md

```diff
@@ -672,242 +679,366 @@ independently.
 
-### Aggregate Share Request {#aggregate-share-request}
+Once the helper has processed each report share in `AggregateInitReq.seq`, the helper
+then creates an AggregateInitResp message to complete its initialization. This message is
+structured as follows:
 
-Once the aggregators have verified at least as many reports as required for the
-PPM task, the leader issues an "aggregate-share request" to each helper. The
-helper responds to this request by extracting its aggregate share from its state
-and encrypting it under the collector's HPKE public key.
+~~~
+enum {
+  continued(0),
+  finished(1)
+  failed(2),
+} ProcessType;
 
-[OPEN ISSUE: consider updating the `checksum` algorithm to not permit
-collisions]
+struct {
+  Nonce nonce;
+  ProcessType process_type;
+  select (ProcessShare.process_type) {
+    case continued: opaque agg_msg<0..2^16-1>; // VDAF output message
+    case finished:  // empty
+    case failed:    ReportShareError;
+  }
+} ProcessShare;
 
-First, the leader computes a checksum over the set of output shares included in
-the batch window. The checksum is computed by taking the SHA256 hash of each
-nonce from the client reports included in the aggregation, then combining the
-hash values with a bitwise-XOR operation.
+struct {
+  opaque helper_state<0..2^16>;
+  ProcessShare seq<1..2^16-1>;
+} AggregateInitResp;
+~~~
 
-Then, for each aggregator endpoint `[aggregator]` in the parameters associated
-with `CollectReq.task_id` (see {{pa-collect}}) except its own, the leader sends
-a POST request to `[aggregator]/aggregate_share` with the following message:
+The `helper_state` parameter contains the helper's initial state. This is an
+optional parameter of an aggregate request that the helper can use to carry
+state across requests. At least part of the state will usually need to be
+encrypted in order to protect user privacy. However, the details of precisely
+how the state is encrypted and the information that it carries is up to the
+helper implementation.
+
+[[OPEN ISSUE: we may end up removing helper_state. See #185]]
+
+The rest of the message is a sequence of ProcessShare values, the order of which
+matches that of the InputReportShare values in `AggregateInitReq.seq`. Each report
+that was marked as invalid is assigned the ProcessType `failed`. Otherwise, the
+ProcessShare is either marked as continued with the output `agg_msg`, or is marked
+as finished if the VDAF computation is finished.
+
+The helper's response to the leader is an HTTP 200 OK whose body is the
+AggregateInitResp and media type is "message/ppm-aggregate-init-resp".
+
+Upon receipt of a helper's AggregateInitResp message, the leader checks that the
+sequence of ProcessShare messages corresponds to the ReportShare sequence of the
+AggregateInitReq. If any message appears out of order, is missing, has an
+unrecognized nonce, or if two messages have the same nonce, then the leader MUST
+abort with error "unrecognizedMessage".
```

Comment:
Why are we both having order and the nonce? Seems like a great opportunity for things to go wrong.

Reply:
I think we can probably relax this. Let's address it in a followup issue.

Reply:
Followup issue: https://github.com/ietf-wg-ppm/ppm-specification/issues/217 (FWIW, I agree this can be relaxed, I included some potential considerations in the issue.)